### DISCUSSION TITLE
Feature request: E2 Per player limits

### DISCUSSION THREAD

**Discussion by wrefgtzweve** [wiremod/wire#2764] [Suggestions]
Currently E2 doesn't have per player limits like SF, this means that limiting E2's on larger servers is incredibly annoying as you either give people few chips or a bunch of low quota chips which both will break peoples builds. So currently i have to choose between low single chip limits which breaks a lot of badly coded public E2's which sadly a large amount of people use or i lower the chip count which breaks a lot of long term builders their stuff as they tend to use lots of low powered chips.

Here's how SF does it <https://github.com/thegrb93/StarfallEx/blob/ba765969401b4171ad22157b9a0f6ab69ac56dba/lua/starfall/instance.lua#L755>

**Comment by Omalike1** [comment]
THIS PLEASE

**Comment by pekmezkruh** [comment]
please add this bro

**Comment by Worldproto** [comment]
add this pls

**Comment by Omalike1** [comment]
If god is real he should send Jesus back down to add this to e2.

**Comment by snepderg** [comment]
This please

**Comment by vurvdev** [comment]
It's not so simple to implement this, it's yet another internal component exposed to addons that can be messed up pretty easily.

Ops are just stored per E2 instance, which is why it's so simple to just do `self.prf = self.prf + 5` or something.
Even functions internally do `if self.prf > e2_tickquota then error("perf") end` (not future proof but pretty ideal for performance)

So implementing this would not only break a few addons, but would potentially be a performance hit with how ops are calculated.

For starfall, it's easy cause:
1. All they have to deal with is cpu usage.
2. A debug hook handles infinite loops, so all they need to worry about is the amount of rolling cpu usage after a tick.

## Conclusion

I strongly advise nobody try to implement this at the very least until the compiler rewrite is merged.
Need a well thought out implementation.

**Reply by wrefgtzweve** [reply]
Could relatively easily be added to the timequota, its in a single spot and highly likely wont interfere with any addons / rewrite

https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/init.lua#L295

**Reply by vurvdev** [reply]
Time quota is not functional and disabled by default, so I don't think this would help much

**Reply by wrefgtzweve** [reply]
I've tested it, its functional. It works pretty well as it limits max cpu rather than ops, limiting it by that would be effectively the same as starfalls limit.

**Reply by vurvdev** [reply]
It's functional but not completely to the point of being enabled by default. I know on TAS, having enabled it accidentally, a bunch of chips end up randomly hitting quota. Or maybe that was just bad config. Can ask @thegrb93 why it's disabled by default

**Reply by Divran** [reply]
One of the reasons the time quota can't be enabled by default is that if the server happens to randomly have a lag spike during an e2 execution, it will falsely be attributed to the e2 and may make it hit the quota and crash.

**Reply by thegrb93** [reply]
@Vurv78 plebs complain if its on
